### PR TITLE
Create topology Python wrapper

### PR DESCRIPTION
**Pull request by jorgensd**
Wrap Topology as a Python class.

Follows along lines #3403

### CODE REVIEW

File: python/dolfinx/mesh.py [hunk 1]
```diff
@@ -67,10 +66,140 @@
 ]
 
 
+class Topology:
+    """Topology for for a :class:`dolfinx.mesh.Mesh`"""
```
Comment (RemDelaporteMathurin) [inline python/dolfinx/mesh.py:70]:
```suggestion
    """Topology for a :class:`dolfinx.mesh.Mesh`"""
```
```suggestion
    """Topology for for a :class:`dolfinx.mesh.Mesh`"""
```
```suggestion
    """Topology for for a :class:`dolfinx.mesh.Mesh`"""
```

File: python/dolfinx/mesh.py [hunk 2]
```diff
@@ -67,10 +66,140 @@
...
+    def dim(self) -> int:
+        """Return the topological dimension of the mesh."""
+        return self._cpp_object.dim
+
+    @property
+    def entity_types(self) -> list[list[CellType]]:
+        """Get the entity types in the topology for all topological dimensions."""
+        return self._cpp_object.entity_types
+
+    def get_cell_permutation_info(self) -> npt.NDArray[np.uint32]:
```
Comment (jhale) [inline python/dolfinx/mesh.py:142]: The content of each np.uint32 being interpreted as a bitfield should probably be mentioned.

I can only find https://numpy.org/doc/stable/reference/generated/numpy.unpackbits.html in numpy which is for uint8 only.

File: python/dolfinx/mesh.py [hunk 3]
```diff
@@ -67,10 +66,140 @@
...
+        """Returns the permutation information"""
+        return self._cpp_object.get_cell_permutation_info()
+
+    def get_facet_permutations(self) -> npt.NDArray[np.uint8]:
```
Comment (jhale) [inline python/dolfinx/mesh.py:146]: The content of each np.uint8 being interpreted as a bitfield should probably be mentioned along with a pointer to `np.unpackbits`.

File: python/dolfinx/mesh.py [hunk 4]
```diff
@@ -67,10 +66,140 @@
...
+        """Get the permutation number to apply to a facet."""
```
Comment (jhale) [inline python/dolfinx/mesh.py:147]: a facet? Isn't this all of the facets?

File: python/test/unit/mesh/test_mixed_topology.py [hunk 5]
```diff
@@ -242,7 +242,7 @@ def test_create_entities():
 
     quad_v = mesh.topology.connectivity((2, qi), (0, 0))
     tri_v = mesh.topology.connectivity((2, ti), (0, 0))
-    ims = mesh.topology.index_maps(2)
+    ims = mesh.topology._cpp_object.index_maps(2)
```
Comment (jhale) [inline python/test/unit/mesh/test_mixed_topology.py:245]: Should this go through the new or extended version of the new interface?

File: python/dolfinx/mesh.py [hunk 6]
```diff
@@ -67,10 +66,140 @@
...
+        Args: The C++ topology object
+        Note:
+            topology objects should not usually be created using this
+            initializer directly.
+        """
+        self._cpp_object = topology
+
+    def cell_name(self) -> str:
+        """String representation of the cell-type of the topology"""
+        return self._cpp_object.cell_name()
```
Comment (jhale) [inline python/dolfinx/mesh.py:85]: Is this necessary? Can't it come from the `CellType` itself?

File: python/dolfinx/mesh.py [hunk 7]
```diff
@@ -67,10 +66,140 @@
...
+
+    @property
+    def comm(self):
+        return self._cpp_object.comm
+
+    def create_connectivity(self, d0: int, d1: int):
+        """Create connectivity between given pair of dimensions, ``d0`` and  ``d1``.
+
+        Args:
+            d0: Dimension of entity one is mapping from
```
Comment (jhale) [inline python/dolfinx/mesh.py:112]: plural entities?

File: python/dolfinx/mesh.py [hunk 8]
```diff
@@ -67,10 +66,140 @@
...
+            d1: Dimension of entity one is mapping to
```
Comment (jhale) [inline python/dolfinx/mesh.py:113]: plural entities?

File: python/dolfinx/mesh.py [hunk 9]
```diff
@@ -67,10 +66,140 @@
...
+        self._cpp_object.create_connectivity(d0, d1)
+
+    def create_entities(self, dim: int) -> int:
+        """Create entities of given topological dimension.
+
+        Args:
+            dim: Topological dimension
+
+        Returns:
+            Number of newly created entities, returns -1 if entities already existed
```
Comment (jhale) [inline python/dolfinx/mesh.py:124]: `-1` as sentinel doesn't seem very Pythonic - `None`?

File: python/dolfinx/mesh.py [hunk 10]
```diff
@@ -67,10 +66,140 @@
...
+        """
+        return self._cpp_object.create_entities(dim)
+
+    def create_entity_permutations(self):
```
Comment (jhale) [inline python/dolfinx/mesh.py:128]: Possible to type return argument?

File: python/dolfinx/mesh.py [hunk 11]
```diff
@@ -67,10 +66,140 @@
...
+
+        Args:
+            d0: Dimension of entity one is mapping from
+            d1: Dimension of entity one is mapping to
+        """
+        if (conn := self._cpp_object.connectivity(d0, d1)) is not None:
+            return conn
+        else:
+            raise RuntimeError(
+                f"Connectiivty between dimension {d0} and {d1} has not been computed."
```
Comment (jorgensd) [inline python/dolfinx/mesh.py:101]:
```suggestion
                f"Connectivity between dimension {d0} and {d1} has not been computed."
```

File: python/dolfinx/mesh.py (hunk 8 above)
Comment (jorgensd) [inline python/dolfinx/mesh.py:113]:
```suggestion
            d0: Dimension of entities one is mapping from
            d1: Dimension of entities one is mapping to
```

File: python/dolfinx/mesh.py (hunk 9 above)
Comment (jorgensd) [reply python/dolfinx/mesh.py:124]: I think it is overkill to check for `-1` and convert, as the return-type is rarely (if ever) used by anyone.

I didn't even know about it until I wrapped it up

File: python/dolfinx/mesh.py (hunk 10 above)
Comment (jorgensd) [reply python/dolfinx/mesh.py:128]: No, there is no return type in C++: https://github.com/FEniCS/dolfinx/blob/202b446cd4a79a1545bfd750e24fa54b6f5f6876/cpp/dolfinx/mesh/Topology.h#L190

File: python/dolfinx/mesh.py (hunk 2 above)
Comment (jorgensd) [reply python/dolfinx/mesh.py:142]: Yes, there doesn't seem to be a way of unpacking nicely:
https://stackoverflow.com/questions/18296035/how-to-extract-the-bits-of-larger-numeric-numpy-data-types

File: python/dolfinx/mesh.py (hunk 3 above)
Comment (jorgensd) [reply python/dolfinx/mesh.py:146]: Do we really need to mention this?  Will anyone ever unpack the permutations (except maybe @mscroggs?)

Comment (jhale) [reply python/dolfinx/mesh.py:145]: If it's in the interface, I'd like to know how to use it.

File: python/dolfinx/mesh.py [hunk 12]
```diff
@@ -65,10 +64,141 @@
...
+class Topology:
+    """Topology for a :class:`dolfinx.mesh.Mesh`"""
+
+    _cpp_object: _cpp.mesh.Topology
+
+    def __init__(self, topology: _cpp.mesh.Topology):
+        """Initialize a topology from a C++ topology.
+        Args: The C++ topology object
+        Note:
+            topology objects should not usually be created using this
```
Comment (jhale) [inline python/dolfinx/mesh.py:76]: Also leave an explicit pointer to the constructor here.

File: python/dolfinx/mesh.py (hunk 9 above)
Comment (jhale) [reply python/dolfinx/mesh.py:123]: OK. I think we will switch to using std::optional instead of sentinel values - we can handle this then.

**Review by jhale — APPROVED** [review]

File: python/dolfinx/mesh.py (hunk 12 above)
Comment (jorgensd) [reply python/dolfinx/mesh.py]: We only have a C++ binding for a constructor atm, so leaving that here for now.

File: python/dolfinx/mesh.py (hunk 3 above)
Comment (jorgensd) [reply python/dolfinx/mesh.py:150]: Ive made a relevant issue #3407 and tried to amend this docstring. let me knowif it is sufficient.

File: python/dolfinx/mesh.py [hunk 13]
```diff
@@ -65,10 +64,156 @@
...
+        Args:
+            dim: Topological dimension
+
+        Returns:
+            Index map for the entities of dimension `dim`.
+        """
+        if (imap := self._cpp_object.index_map(dim)) is not None:
+            return imap
+        else:
+            raise RuntimeError(f"Entities of dimension {dim} has not been computed")
```
Comment (jhale) [inline python/dolfinx/mesh.py:176]: Instruct user how to fix the issue.

File: python/dolfinx/mesh.py [hunk 14]
```diff
@@ -65,10 +64,156 @@
...
+    @property
+    def entity_types(self) -> list[list[CellType]]:
+        """Get the entity types in the topology for all topological dimensions."""
+        return self._cpp_object.entity_types
+
+    def get_cell_permutation_info(self) -> npt.NDArray[np.uint32]:
+        """Returns the permutation information
+
+        The returned data is used for packing coefficients and assembling of tensors.
+        Each integer encodes the number of reflections and permutations for each sub-entity
```
Comment (jhale) [inline python/dolfinx/mesh.py:145]: The bits of each integer

File: python/dolfinx/mesh.py [hunk 15]
```diff
@@ -65,10 +64,156 @@
...
+        of the cell to be able to map it to the reference element.
+        """
+        return self._cpp_object.get_cell_permutation_info()
+
+    def get_facet_permutations(self) -> npt.NDArray[np.uint8]:
+        """Get the permutation number to apply to facets.
+
+        An entry describes the number of reflections and rotations that has to
```
Comment (jhale) [inline python/dolfinx/mesh.py:153]: The bits of each integer

**Review by jhale — CHANGES_REQUESTED** [review]

File: python/dolfinx/mesh.py [hunk 16]
```diff
@@ -65,10 +64,159 @@
...
+
+    def cell_name(self) -> str:
+        """String representation of the cell-type of the topology"""
+        return to_string(self._cpp_object.cell_type)
+
+    def connectivity(self, d0: int, d1: int) -> _cpp.graph.AdjacencyList_int32:
+        """Return connectivity from entities of dimension ``d0`` to entities of dimension ``d1``.
+
+        Note:
+            Assumes only one entity type per dimension.
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:89]: I don't understand this note.

Comment (jorgensd) [reply python/dolfinx/mesh.py:89]: This is taken directly from the C++ docs:
https://github.com/FEniCS/dolfinx/blob/main/cpp/dolfinx/mesh/Topology.h#L106-L130
and I interpret this as:
This connectivity does not take into account prism meshes or mixed meshes, as you would have multiple cell types per dimension. @chrisrichardson Feel free to correct me if I am wrong.

**Review by jhale — APPROVED** [review]

File: python/dolfinx/mesh.py [hunk 17]
```diff
@@ -65,10 +64,159 @@
...
+        return self._cpp_object.get_facet_permutations()
+
+    def index_map(self, dim: int) -> _cpp.common.IndexMap:
+        """Get the IndexMap that described the parallel distribution of the mesh entities.
+
+        Args:
+            dim: Topological dimension
+
+        Returns:
+            Index map for the entities of dimension `dim`.
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:171]:
```suggestion
            Index map for the entities of dimension ``dim``.
```

File: python/dolfinx/mesh.py (hunk 16 above)
Comment (jhale) [reply python/dolfinx/mesh.py:92]: From a user's point of view the important point would be to mention that this connectivity is not correct for meshes containing prisms and/or mixed meshes.

Comment (garth-wells) [reply python/dolfinx/mesh.py:92]: Let's just remove the comment then. At this stage, anyone using mixed cell types is on their own . . .

File: python/dolfinx/mesh.py [hunk 18]
```diff
@@ -68,6 +67,154 @@
...
+
+    def cell_name(self) -> str:
+        """String representation of the cell-type of the topology"""
+        return to_string(self._cpp_object.cell_type)
+
+    def connectivity(self, d0: int, d1: int) -> _cpp.graph.AdjacencyList_int32:
+        """Return connectivity from entities of dimension ``d0`` to entities of dimension ``d1``.
+
+        Note:
+            Assumes only one entity type per dimension.
```
Comment (jorgensd) [inline python/dolfinx/mesh.py:92]:
```suggestion
```

**Review by garth-wells — APPROVED** [review]

**Pull request merged**